apollo-router/src/plugins/telemetry/metrics/apollo.rs

```diff
@@ -326,55 +337,66 @@ impl ApolloMetricsExporter {
     }
 }
 
-fn aggregate(
-    metrics: Vec<Metrics>,
-) -> Vec<(
-    String,
-    ContextualizedStats,
-    HashMap<String, ReferencedFieldsForType>,
-    u64,
-)> {
+fn aggregate(metrics: Vec<Metrics>) -> HashMap<MetricsKey, AggregatedMetrics> {
     let mut aggregated_metrics = HashMap::<_, AggregatedMetrics>::new();
-
     for metric in metrics {
-        *aggregated_metrics
-            .entry(
-                (
-                    metric.client_name.clone(),
-                    metric.client_version.clone(),
-                    metric.stats_report_key.clone(),
-                )
-                    .clone(),
-            )
-            .or_default() += metric;
+        *aggregated_metrics.entry(metric.key.clone()).or_default() += metric;
     }
-
     aggregated_metrics
-        .into_iter()
-        .map(|((client_name, client_version, key), aggregated_metric)| {
-            let mut contextualized_stats = ContextualizedStats {
-                context: Some(StatsContext {
-                    client_name,
-                    client_version,
-                }),
-                ..Default::default()
-            };
-            contextualized_stats.per_type_stat = aggregated_metric
-                .per_type_stat
-                .into_iter()
-                .map(|(k, v)| (k, v.into()))
-                .collect();
-            contextualized_stats.query_latency_stats =
-                Some(aggregated_metric.query_latency_stats.into());
-
-            (
-                key,
-                contextualized_stats,
-                aggregated_metric.referenced_fields_by_type,
-                aggregated_metric.operation_count,
-            )
-        })
-        .collect()
+}
+
+fn to_report(
+    header: ReportHeader,
+    aggregated_metrics: HashMap<MetricsKey, AggregatedMetrics>,
+) -> apollo_spaceport::Report {
+    let time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap_or_default();
+    let seconds = time.as_secs();
+    let nanos = time.as_nanos() - (seconds as u128 * 1_000_000_000);
```

Comment:
There's a subsec_nanos call you can use here... which I learned from reading the source of https://docs.rs/prost-types/latest/prost_types/struct.Timestamp.html#impl-From%3CSystemTime%3E which suggests that you should be able to go directly from SystemTime to Timestamp without needing to write the conversion code yourself.

Reply:
Nice, fixed in [4e8e3b4](https://github.com/apollographql/router/pull/1020/commits/4e8e3b43aa7046a1bbfb325798305c83880fd499)